aten/src/ATen/native/cpu/ActivationKernel.cpp

```diff
@@ -0,0 +1,484 @@
+    int64_t leftover = _leftover<Vec::size>(d, dim_size);
+    Vec value;
+    value.load(input_data + d, leftover);
+    if (leftover < Vec::size) {
+      for (int i = 0; i < leftover; i++) {
+        if (input_data[d + i] > max_val) {
+          max_val = input_data[d + i];
+        }
+      }
+    } else {
+      max_val_vec = vec256::max(value, max_val_vec);
+    }
+  }
+  scalar_t max_val_arr[Vec::size];
+  max_val_vec.store(max_val_arr);
+  for (int64_t i = 0; i < Vec::size; i++) {
+    if (max_val_arr[i] > max_val) {
+      max_val = max_val_arr[i];
+    }
+  }
+  return max_val;
+}
+
+template <typename scalar_t>
+inline void _vec_mul_scalarsub_write(
+    scalar_t* grad_input_data,
+    scalar_t* grad_data,
+    scalar_t* output_data,
+    int64_t dim_size,
+    scalar_t sum) {
+  using Vec = vec256::Vec256<scalar_t>;
+  Vec sum_vec(sum);
+  int64_t d = 0;
+  for (; d < dim_size; d += Vec::size) {
+    Vec output(0);
+    Vec grad(0);
+    int64_t leftover = _leftover<Vec::size>(d, dim_size);
+    output.load(output_data + d, leftover);
+    grad.load(grad_data + d, leftover);
+    grad = grad - sum_vec;
+    grad = grad * output;
+
+    grad.store(grad_input_data + d, leftover);
+  }
+}
+
+template <typename scalar_t>
+inline void _vec_sub_exp_scalarmul_write(
+    scalar_t* grad_input_data,
+    scalar_t* grad_data,
+    scalar_t* output_data,
+    int64_t dim_size,
+    scalar_t sum) {
+  using Vec = vec256::Vec256<scalar_t>;
+  Vec sum_vec(sum);
+  int64_t d = 0;
+  for (; d < dim_size; d += Vec::size) {
+    Vec output(0);
+    Vec grad(0);
+    int64_t leftover = _leftover<Vec::size>(d, dim_size);
+    output.load(output_data + d, leftover);
+    grad.load(grad_data + d, leftover);
+    output = output.exp();
+    output = output * sum_vec;
+    grad = grad - output;
+
+    grad.store(grad_input_data + d, leftover);
+  }
+}
+
+template <typename scalar_t>
+inline scalar_t _vec_norm_exp_sum_write(
+    scalar_t* output_data,
+    int64_t dim_size,
+    scalar_t* input_data,
+    scalar_t max_input) {
+  using Vec = vec256::Vec256<scalar_t>;
+
+  int64_t d = 0;
+  Vec tmpsum(0);
+  Vec max_input_vec(max_input);
+  scalar_t tmp_sum_arr[Vec::size];
+  scalar_t tmp_sum_scalar = 0;
+  for (; d < dim_size; d += Vec::size) {
+    int64_t leftover = _leftover<Vec::size>(d, dim_size);
+    Vec value(0);
+    value.load(input_data + d, leftover);
+    value = value - max_input_vec;
+    value = value.exp();
+    // Need to do a partial add
+    if (d + Vec::size > dim_size) {
+      scalar_t value_arr[Vec::size];
+      value.store(value_arr);
+      for (int64_t i = 0; i < dim_size - d; i++) {
+        tmp_sum_scalar += value_arr[i];
+      }
+      value.store(output_data + d, leftover);
+    } else {
+      tmpsum = tmpsum + value;
+      value.store(output_data + d);
+    }
+  }
+  tmpsum.store(tmp_sum_arr);
+  for (int64_t i = 0; i < Vec::size; i++) {
+    tmp_sum_scalar += tmp_sum_arr[i];
+  }
+  return tmp_sum_scalar;
+}
```

Comment:
TBH all of the 6 functions above are variants of something similar to `tensor_apply`, except it's a reduce done on pointers of a given size (so it's more like `vec_reduce`). They do a map operation on multiple parallel inputs, and generally have the same body modulo an operation that's happening in the inner loop + the reduction of the vector at the end. Having a generic function for this would likely make the code a lot shorter and simpler.

Reply:
Agreed.